include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -673,241 +500,442 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
-	if (op==OP_DIFFTRANS) {
-		instr_execute_cos_eval_DIFFTRANS(instr, regs, params, bsdf_data);
-	} else
+			
+		bxdf_eval = is_valid ? vec3(bxdf_eval_scalar_part) : vec3(0.0);
+
+#ifdef OP_CONDUCTOR
+		if (op == OP_CONDUCTOR)
+			bxdf_eval *= fr;
+		else
 #endif
-#ifdef OP_DIELECTRIC
-	if (op==OP_DIELECTRIC) {
-		instr_execute_cos_eval_DIELECTRIC(instr, regs, params, bsdf_data);
+		{
+			float reflectance = dot(CIE_XYZ_Luma_Y_coeffs, fr);
+			reflectance = refraction ? (1.0 - reflectance) : reflectance;
+			pdf *= reflectance;
+			bxdf_eval *= reflectance;
+		}
 	} else
 #endif
-#ifdef OP_THINDIELECTRIC
-	if (op==OP_THINDIELECTRIC) {
-		instr_execute_cos_eval_THINDIELECTRIC(instr, regs, params, bsdf_data);
-	} else
+	{}
+
+	uvec3 regs = instr_decodeRegisters(instr);
+
+	eval_and_pdf_t result = eval_and_pdf_t(bxdf_eval, pdf);
+	if (!op_isBXDF(op))
+	{
+		mat2x4 srcs = instr_fetchSrcRegs(instr, regs);
+		BEGIN_CASES(op)
+#ifdef OP_COATING
+		CASE_BEGIN(op, OP_COATING) {
+			result = instr_execute_cos_eval_pdf_COATING(instr, srcs, params, ior[0], ior2[0], s, microfacet, bsdf_data);
+		} CASE_END
 #endif
 #ifdef OP_BLEND
-	if (op==OP_BLEND) {
-		pdf = instr_execute_cos_eval_pdf_BLEND(instr, regs, params, bsdf_data);
-	} else
+		CASE_BEGIN(op, OP_BLEND) {
+			result = instr_execute_cos_eval_pdf_BLEND(instr, srcs, params, bsdf_data);
+		} CASE_END
 #endif
 #ifdef OP_BUMPMAP
-	if (op==OP_BUMPMAP) {
-		instr_execute_BUMPMAP(instr, L); // @Crisspl is your L a global or something?
-	} else
+		CASE_BEGIN(op, OP_BUMPMAP) {
+			instr_execute_BUMPMAP(instr, srcs, precomp);
+		} CASE_END
 #endif
 #ifdef OP_SET_GEOM_NORMAL
-	if (op==OP_SET_GEOM_NORMAL) {
-		instr_execute_SET_GEOM_NORMAL(L); // @Crisspl is your L a global or something?
-	} else
+		CASE_BEGIN(op, OP_SET_GEOM_NORMAL) {
+			instr_execute_SET_GEOM_NORMAL(instr, precomp);
+		} CASE_END
 #endif
-	{} //else "empty braces"
+		CASE_OTHERWISE
+		{} //else "empty braces"
+		END_CASES
+	}
 
-	if (op_isBXDForBlend(op)) //OP_BLEND writes pdf as well, but it happens in instr_execute_cos_eval_pdf_BLEND()
-		writeReg(REG_DST(regs)+3u, pdf);
+	if (op_isBXDForCoatOrBlend(op))
+		writeReg(REG_DST(regs), result);
 }
 
-vec3 irr_bsdf_cos_remainder_and_pdf(in instr_stream_t stream, in irr_glsl_BSDFSample s, out float _out_pdf)
+eval_and_pdf_t irr_bsdf_eval_and_pdf(in MC_precomputed_t precomp, in instr_stream_t stream, in uint generator_offset, inout irr_glsl_LightSample s, inout irr_glsl_AnisotropicMicrofacetCache microfacet)
 {
-#ifndef NO_TWOSIDED
-	bool ts = false;
-#endif
+	setCurrInteraction(precomp);
 	for (uint i = 0u; i < stream.count; ++i)
 	{
 		instr_t instr = irr_glsl_MC_fetchInstr(stream.offset+i);
-#ifndef NO_TWOSIDED
-		handleTwosided(ts, instr);
+		uint op = instr_getOpcode(instr);
+
+		bool skip = (i == generator_offset);
+		// skip deltas
+#ifdef OP_THINDIELECTRIC
+		skip = skip || (op == OP_THINDIELECTRIC);
+#endif
+#ifdef OP_DELTATRANS
+		skip = skip || (op == OP_DELTATRANS);
+#endif
+
+		if (skip)
```

Comment:
pass the skip into `instr_eval_and_pdf_execute` you have divergence for doing the exact same thing (decoding registers and writing the result)

Reply:
and remember to init `pdf = skip ? 0.0:1.0;`